lib/jsdom/living/xhr/XMLHttpRequest-impl.js

```diff
@@ -0,0 +1,1109 @@
+      .map(key => [conversions.ByteString(key).toLowerCase(), properties.responseHeaders[key]].join(": "))
+      .join("\r\n");
+  }
+
+  /** @param {string} header */
+  getResponseHeader(header) {
+    const { properties, readyState } = this;
+    if (readyState === XMLHttpRequest.UNSENT || readyState === XMLHttpRequest.OPENED) {
+      return null;
+    }
+    const lcHeader = conversions.ByteString(header).toLowerCase();
+    if (properties.filteredResponseHeaders.find(filtered => lcHeader === filtered.toLowerCase())) {
+      return null;
+    }
+    return getResponseHeader(this, lcHeader);
+  }
+
+  /**
+   * @param {string} method
+   * @param {string} uri
+   * @param {boolean} [asynchronous]
+   * @param {string | null} [user]
+   * @param {string | null} [password]
+   */
+  open(method, uri, asynchronous, user, password) {
+    if (!this._ownerDocument) {
+      throw DOMException.create(this._globalObject, ["The object is in an invalid state.", "InvalidStateError"]);
+    }
+    const { flag, properties } = this;
+
+    if (!tokenRegexp.test(method)) {
+      throw DOMException.create(this._globalObject, ["The string did not match the expected pattern.", "SyntaxError"]);
+    }
+    const upperCaseMethod = method.toUpperCase();
+    if (forbiddenRequestMethods.has(upperCaseMethod)) {
+      throw DOMException.create(this._globalObject, ["The operation is insecure.", "SecurityError"]);
+    }
+
+    const { client } = properties;
+    if (client && typeof client.abort === "function") {
+      client.abort();
+    }
+
+    if (allowedRequestMethods.has(upperCaseMethod)) {
+      method = upperCaseMethod;
+    }
+    if (typeof asynchronous !== "undefined") {
+      flag.synchronous = !asynchronous;
+    } else {
+      flag.synchronous = false;
+    }
+    if (flag.responseType && flag.synchronous) {
+      throw DOMException.create(this._globalObject, [
+        "The object does not support the operation or argument.",
+        "InvalidAccessError"
+      ]);
+    }
+    if (flag.synchronous && flag.timeout) {
+      throw DOMException.create(this._globalObject, [
+        "The object does not support the operation or argument.",
+        "InvalidAccessError"
+      ]);
+    }
+    flag.method = method;
+
+    let urlObj;
+    try {
+      urlObj = new URL(uri, documentBaseURLSerialized(this._ownerDocument));
+    } catch (e) {
+      throw DOMException.create(this._globalObject, ["The string did not match the expected pattern.", "SyntaxError"]);
+    }
+
+    if (user || (password && !urlObj.username)) {
+      flag.auth = {
+        user,
+        pass: password
+      };
+      urlObj.username = "";
+      urlObj.password = "";
+    }
+
+    flag.uri = urlObj.href;
+    flag.requestHeaders = {};
+    flag.preflight = false;
+
+    properties.send = false;
+    properties.uploadListener = false;
+    properties.abortError = false;
+    this.responseURL = "";
```

Comment:
The original also resets `properties.requestBuffer` and `properties.requestCache`. Is that unnecessary?